File: tailor-cloud-backend/internal/config/domain/ambassador.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// Ambassador アンバサダーモデル
// Phase 1: 学生アンバサダーが「誰が売ったか」を記録し、成果報酬を自動計算する基盤
type Ambassador struct {
	ID          string    `json:"id" db:"id"`
	TenantID    string    `json:"tenant_id" db:"tenant_id"`
	UserID      string    `json:"user_id" db:"user_id"` // Firebase AuthのUserIDとリンク
	Name        string    `json:"name" db:"name"`
	Email       string    `json:"email" db:"email"`
	Phone       string    `json:"phone" db:"phone"`
	Status      AmbassadorStatus `json:"status" db:"status"`
	CommissionRate float64 `json:"commission_rate" db:"commission_rate"` // 成果報酬率（例: 0.10 = 10%）
	TotalSales  int64     `json:"total_sales" db:"total_sales"` // 累計売上（円）
	TotalCommission int64 `json:"total_commission" db:"total_commission"` // 累計報酬（円）
	CreatedAt   time.Time `json:"created_at" db:"created_at"`
	UpdatedAt   time.Time `json:"updated_at" db:"updated_at"`
}

// AmbassadorStatus アンバサダーステータス
type AmbassadorStatus string

const (
	AmbassadorStatusActive   AmbassadorStatus = "Active"   // アクティブ
	AmbassadorStatusInactive AmbassadorStatus = "Inactive" // 非アクティブ
	AmbassadorStatusSuspended AmbassadorStatus = "Suspended" // 一時停止
)

// Commission 成果報酬モデル
// 注文ごとの成果報酬を記録
type Commission struct {
	ID            string    `json:"id" db:"id"`
	OrderID       string    `json:"order_id" db:"order_id"`
	AmbassadorID  string    `json:"ambassador_id" db:"ambassador_id"`
	TenantID      string    `json:"tenant_id" db:"tenant_id"`
	OrderAmount   int64     `json:"order_amount" db:"order_amount"` // 注文金額
	CommissionRate float64  `json:"commission_rate" db:"commission_rate"` // 報酬率
	CommissionAmount int64  `json:"commission_amount" db:"commission_amount"` // 報酬額（円）
	Status        CommissionStatus `json:"status" db:"status"`
	PaidAt        *time.Time `json:"paid_at" db:"paid_at"` // 支払日（nil = 未払い）
	CreatedAt     time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt     time.Time  `json:"updated_at" db:"updated_at"`
}

// CommissionStatus 成果報酬ステータス
type CommissionStatus string

const (
	CommissionStatusPending   CommissionStatus = "Pending"   // 未確定（注文が確定していない）
	CommissionStatusApproved  CommissionStatus = "Approved"  // 確定（支払い待ち）
	CommissionStatusPaid      CommissionStatus = "Paid"      // 支払済み
	CommissionStatusCancelled CommissionStatus = "Cancelled" // キャンセル（注文がキャンセルされた場合）
)

// CalculateCommission 成果報酬を計算
func CalculateCommission(orderAmount int64, commissionRate float64) int64 {
	return int64(float64(orderAmount) * commissionRate)
}

// NewAmbassador 新しいアンバサダーを作成
func NewAmbassador(tenantID, userID, name, email string) *Ambassador {
	now := time.Now()
	
	// デフォルトの成果報酬率: 10%（Phase 1では固定値）
	defaultCommissionRate := 0.10
	
	return &Ambassador{
		ID:            uuid.New().String(),
		TenantID:      tenantID,
		UserID:        userID,
		Name:          name,
		Email:         email,
		Status:        AmbassadorStatusActive,
		CommissionRate: defaultCommissionRate,
		TotalSales:    0,
		TotalCommission: 0,
		CreatedAt:     now,
		UpdatedAt:     now,
	}
}

// NewCommission 新しい成果報酬を作成
func NewCommission(orderID, ambassadorID, tenantID string, orderAmount int64, commissionRate float64) *Commission {
	commissionAmount := CalculateCommission(orderAmount, commissionRate)
	
	return &Commission{
		ID:              uuid.New().String(),
		OrderID:         orderID,
		AmbassadorID:    ambassadorID,
		TenantID:        tenantID,
		OrderAmount:     orderAmount,
		CommissionRate:  commissionRate,
		CommissionAmount: commissionAmount,
		Status:          CommissionStatusPending, // 注文確定までPending
		CreatedAt:       time.Now(),
		UpdatedAt:       time.Now(),
	}
}

File: tailor-cloud-app/lib/models/ambassador.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'ambassador.freezed.dart';
part 'ambassador.g.dart';

/// アンバサダーステータス
enum AmbassadorStatus {
  @JsonValue('Active')
  active,
  @JsonValue('Inactive')
  inactive,
  @JsonValue('Suspended')
  suspended,
}

/// 成果報酬ステータス
enum CommissionStatus {
  @JsonValue('Pending')
  pending,
  @JsonValue('Approved')
  approved,
  @JsonValue('Paid')
  paid,
  @JsonValue('Cancelled')
  cancelled,
}

/// アンバサダーモデル
@freezed
class Ambassador with _$Ambassador {
  const factory Ambassador({
    required String id,
    @JsonKey(name: 'tenant_id') required String tenantId,
    @JsonKey(name: 'user_id') required String userId,
    required String name,
    required String email,
    String? phone,
    required AmbassadorStatus status,
    @JsonKey(name: 'commission_rate') required double commissionRate,
    @JsonKey(name: 'total_sales') @Default(0) int totalSales,
    @JsonKey(name: 'total_commission') @Default(0) int totalCommission,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'updated_at') required DateTime updatedAt,
  }) = _Ambassador;

  factory Ambassador.fromJson(Map<String, dynamic> json) =>
      _$AmbassadorFromJson(json);
}

/// 成果報酬モデル
@freezed
class Commission with _$Commission {
  const factory Commission({
    required String id,
    @JsonKey(name: 'order_id') required String orderId,
    @JsonKey(name: 'ambassador_id') required String ambassadorId,
    @JsonKey(name: 'tenant_id') required String tenantId,
    @JsonKey(name: 'order_amount') required int orderAmount,
    @JsonKey(name: 'commission_rate') required double commissionRate,
    @JsonKey(name: 'commission_amount') required int commissionAmount,
    required CommissionStatus status,
    @JsonKey(name: 'paid_at') DateTime? paidAt,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'updated_at') required DateTime updatedAt,
  }) = _Commission;

  factory Commission.fromJson(Map<String, dynamic> json) =>
      _$CommissionFromJson(json);
}

/// アンバサダー拡張メソッド
extension AmbassadorExtension on Ambassador {
  /// 成果報酬率の表示（%）
  String get commissionRateDisplay {
    return '${(commissionRate * 100).toStringAsFixed(1)}%';
  }

  /// 累計売上の表示（円）
  String get totalSalesDisplay {
    return '¥${totalSales.toString().replaceAllMapped(
          RegExp(r'(\d)(?=(\d{3})+(?!\d))'),
          (Match m) => '${m[1]},',
        )}';
  }

  /// 累計報酬の表示（円）
  String get totalCommissionDisplay {
    return '¥${totalCommission.toString().replaceAllMapped(
          RegExp(r'(\d)(?=(\d{3})+(?!\d))'),
          (Match m) => '${m[1]},',
        )}';
  }
}

/// 成果報酬拡張メソッド
extension CommissionExtension on Commission {
  /// 注文金額の表示（円）
  String get orderAmountDisplay {
    return '¥${orderAmount.toString().replaceAllMapped(
          RegExp(r'(\d)(?=(\d{3})+(?!\d))'),
          (Match m) => '${m[1]},',
        )}';
  }

  /// 報酬額の表示（円）
  String get commissionAmountDisplay {
    return '¥${commissionAmount.toString().replaceAllMapped(
          RegExp(r'(\d)(?=(\d{3})+(?!\d))'),
          (Match m) => '${m[1]},',
        )}';
  }

  /// ステータスの表示ラベル
  String get statusLabel {
    switch (status) {
      case CommissionStatus.pending:
        return '未確定';
      case CommissionStatus.approved:
        return '確定';
      case CommissionStatus.paid:
        return '支払済み';
      case CommissionStatus.cancelled:
        return 'キャンセル';
    }
  }
}
